tests/webTables.spec.ts

```diff
@@ -0,0 +1,177 @@
+            if (cellValue == 'Madison') {
+                allMadisonCitizen += 1
+            }
+        }
+        expect(allMadisonCitizen).toEqual(4)
+    })
+
+    test('Validate search by Last Name', async ({ page }) => {
+        const inputLastName = page.locator('#lastName')
+        await inputLastName.fill('Black')
+        const buttonFindOwner = page.getByRole('button', { name: 'Find Owner' })
+        await buttonFindOwner.click()
+
+        await page.waitForTimeout(2000)
+
+        const ownersListLoop = async (searchValue) => {
+            if (searchValue == 'Playwright') {
+                expect(await page.locator('.xd-container div').last().textContent()).toContain(`No owners with LastName starting with "${searchValue}"`)
+                return;
+            }
+
+            const ownersList = page.locator('tbody tr', { has: page.locator('td').nth(2) })
+            for (const row of await ownersList.all()) {
+                const cellValue = await row.locator('td').first().textContent()
+                expect(cellValue).toContain(searchValue)
+
+            }
+        }
+
+        ownersListLoop('Black')
+
+        await inputLastName.clear()
+        await inputLastName.fill('Davis')
+        await buttonFindOwner.click()
+        await page.waitForTimeout(2000)
+        ownersListLoop('Davis')
+
+        await inputLastName.clear()
+        await inputLastName.fill('Es')
+        await buttonFindOwner.click()
+        await page.waitForTimeout(2000)
+        ownersListLoop('Es')
+
+        await inputLastName.clear()
+        await inputLastName.fill('Playwright')
+        await buttonFindOwner.click()
+        await page.waitForTimeout(2000)
+        ownersListLoop('Playwright')
+
+
+    })
+
+    test('Validate phone number and pet name on the Owner Information page', async ({ page }) => {
+        const ownerRow = page.getByRole('row', { name: '6085552765' })
+        const petName = await ownerRow.locator('td').last().textContent()
+        await ownerRow.getByRole('link', { name: 'Peter McTavish' }).click()
+        await page.waitForTimeout(2000)
+        expect(await page.locator('table').first().locator('tr td').last().textContent()).toContain('6085552765')
+        expect(await page.locator('.dl-horizontal dd').first().textContent()).toContain(petName?.trim())
+
+    })
+
+    test('Validate pets of the Madison city', async ({ page }) => {
+        await page.waitForSelector('table')
+
+        const expectedList = [" Leo ", " George ", " Mulligan ", " Freddy "]
+        let petList = []
+
+        const allRows = page.locator('tbody tr', { has: page.locator('td').nth(4) })
+
+        for (const row of await allRows.all()) {
+            const cityValue = await row.locator('td').nth(2).textContent()
+            const petsValue = await row.locator('td').nth(4).textContent()
+
+            if (cityValue == 'Madison') {
+                petList.push(petsValue)
+            }
+        }
```

Comment:
Use the knowledge from this lesson and get all rows of Madison city with just a single locator without a need to loop through all rows of the table to identify those rows.